### ISSUE TITLE
standalone/tls: port test failures because of containers on hosts of OpenVZ guests

### ISSUE THREAD

**Issue by hvisage** [Neilpang/acme.sh#296]
Hi there,

 I'm trying to use the acme.sh on a (ProxMox) host running OpenVZs, which also have SSL and HTTPS ports listening to. This causes failures as the script (using ss/netstat) pickup the OpenVZ container(s) are using port 80/443, and thus it stops and throws an error. 

 Perhaps add an "ignore port test" option?

If that is accetable, I could fork and create a pull request.

**Comment by Neilpang** [comment]
@hvisage  
acme protocol requires to use 80/443 port.   If 80/443 is already used by your openvz container, you can not issue cert then.

I suggest you issue cert in the openvz container, that is using 80 port.

Setting a reverse proxy or use DNS mode may also work for you.

**Comment by hvisage** [comment]
@Neilpang You are 100% correct in the "usual" cases. HOWEVER, only once you've worked with OpenVZ containers will you understand that you are wrong.

What happens is this: the OpenVZ containers share the same process and network space than the host. Ie. when you type netstat -tnlp, you could see *multiple* processes listening on port :80/:443. one on each of the containers that uses those ports, and one for the host that uses that port.

thus this is the output of ss/netstat -tnlp/lsof -i:80 for the host in question:

```
root@HostServer:~# ss -l |grep http
LISTEN     0      128                     *:http                     *:*
LISTEN     0      128                    :::https                   :::*
LISTEN     0      128                     *:https                    *:*
LISTEN     0      128         xxx.yyy.25.19:http                     *:*
LISTEN     0      128         xxx.yyy.25.18:http                     *:*
root@HostServer:~# netstat -tlp |grep http
tcp        0      0 *:http                  *:*                     LISTEN      6786/apache2
tcp        0      0 *:https                 *:*                     LISTEN      6786/apache2
tcp        0      0 server2.org.co:http *:*                     LISTEN      2888/nginx.conf
tcp        0      0 serv.org.co.:http *:*                     LISTEN      2888/nginx.conf
tcp6       0      0 [::]:https              [::]:*                  LISTEN      2998/apache2
root@HostServer:~# lsof -i :80
COMMAND  PID     USER   FD   TYPE  DEVICE SIZE/OFF NODE NAME
nginx   2888     root    8u  IPv4   12936      0t0  TCP serv.org.co.za:http (LISTEN)
nginx   2888     root    9u  IPv4   12937      0t0  TCP serv2.org.co.za:http (LISTEN)
nginx   2889    nginx    7u  IPv4 1108586      0t0  TCP serv.org.co.za:http->rover.umts.vodacom.co.za:32707 (ESTABLISHED)
nginx   2889    nginx    8u  IPv4   12936      0t0  TCP serv.org.co.za:http (LISTEN)
nginx   2889    nginx    9u  IPv4   12937      0t0  TCP serv2.org.co.za:http (LISTEN)
nginx   2889    nginx   14u  IPv4 1108659      0t0  TCP serv.org.co.za:http->crawl-66-249-64-97.googlebot.com:61862 (ESTABLISHED)
apache2 6786     root    3u  IPv4   24613      0t0  TCP *:http (LISTEN)
apache2 6811 www-data    3u  IPv4   24613      0t0  TCP *:http (LISTEN)
apache2 6812 www-data    3u  IPv4   24613      0t0  TCP *:http (LISTEN)
root@HostServer:~#
```

This is actually also shows another thing that comes up where the acme.sh over-simplified test fails: multiple IPs on the same server, while acme.sh should listen on port connected to the Internet, there might still be another service (non-internet based, ie  internal service) that the test will fals on, even though it will work perfectly alright if it bounded to that specific IP(s).

But back to the main problem I've experienced for the over-simplified test that fails with the OpenVZ running:

See, the openvz have apache running on port :80 (different IP than the host) while the Host (serv.org/serv2.org) have nginx listening on port 80. All kosher, 'cause they just appear to be in the same port space, but the OpenVZ/etc. separates the container from the host, even though they show up ;(

The acme.sh (sans the test) will work perfectly on port 80 *if* I shutdown the nginx inside the host, even though I still run the OPenVZ's apache that listens and show to be using port 80.

I know that sounds and is confusing, but believe me, I've been "bitten" by this issue myself where I've been dumbfounded that I don't get a connection on the host even though I see the port being listened on, but it was the OpenVZ container's server that listened on that port, not the host's IP/network.

**Comment by Neilpang** [comment]
@hvisage 

I just add a new branch `addr` for you.

https://github.com/Neilpang/acme.sh/tree/addr

A new parameter `--local-address` is added to specify the listening ip address in standalone mode.

```
acme.sh  --issue   -d  pve.mydomain.com  --standalone  --local-address   78.78.78.78
```

How to install the branch online:
```
export  BRANCH=addr
curl https://get.acme.sh | sh

acme.sh  --issue   -d  pve.mydomain.com  --standalone  --local-address   78.78.78.78

```


Please have a try.    I will merge it to master branch after more testings.

Thanks.

**Comment by Neilpang** [comment]
@hvisage   Merged:  0463b5d6cdc63a69977b89a19791edc560b43d7a
